addon/-private/system/record-map.js

```diff
@@ -0,0 +1,47 @@
+import EmptyObject from 'ember-data/-private/system/empty-object';
+
+export default class ClassMeta {
+  constructor(modelName) {
+    this.modelName = modelName;
+    this._idToRecord = null;
+    this._records = null;
+    this._metadata = null;
+  }
+
+  get idToRecord() {
+    return this._idToRecord || (this._idToRecord = new EmptyObject());
+  }
+
+  get records() {
+    return this._records || (this._records = []);
+  }
+
+  get metadata() {
+    return this._metadata || (this._metadata = new EmptyObject());
+  }
+
+  /* deprecated way of accessing modelClass */
+  get type() {
+    throw new Error('RecordMap.type is no longer available');
+  }
+
+  clear() {
+    if (this._records) {
+      let records = this._records.slice();
```

Comment:
we don't appear to purge `this._records` membership, should we maybe instead do.

```js
let records = this._records;
this._records = []; // or maybe null
```

That way, we release anything currently retained by `this._records` and don't need to copy the `this._records` array via `slice`